### PR TITLE
[PYIC-2574] Store client OAuth session

### PR DESCRIPTION
**Pull request by lwebb-iig**
Update `initialize-ipv-session` lambda to additionally store Client OAuth session in new table

### What changed

- Permissions and envs for DynamoDB table `client-oauth-sessions` to allow `initialize-ipv-session` lambda to read/write to the table
- Creation of client oauth session service to create and write client oauth sessions to dynamo
- Updated ipv session handler to call client oauth session for normal & error jouneys

### Why did it change

Part of refactoring client oauth session management for app journey browser changes

### Issue tracking

- [PYIC-2574](https://govukverify.atlassian.net/browse/PYIC-2574)

### Environment variables or secrets

- [x] No environment variables or secrets were added or changed

### Other considerations

- [x] Update [README](./blob/main/README.md) with any new instructions or tasks

[PYIC-2574]: https://govukverify.atlassian.net/browse/PYIC-2574?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: lambdas/initialise-ipv-session/src/test/java/uk/gov/di/ipv/core/initialiseipvsession/InitialiseIpvSessionHandlerTest.java [hunk 1]
```diff
@@ -34,11 +34,9 @@
 import uk.gov.di.ipv.core.library.exceptions.SqsException;
 import uk.gov.di.ipv.core.library.fixtures.TestFixtures;
 import uk.gov.di.ipv.core.library.helpers.SecureTokenHelper;
+import uk.gov.di.ipv.core.library.persistence.item.ClientOAuthSessionItem;
 import uk.gov.di.ipv.core.library.persistence.item.IpvSessionItem;
-import uk.gov.di.ipv.core.library.service.AuditService;
-import uk.gov.di.ipv.core.library.service.ConfigService;
-import uk.gov.di.ipv.core.library.service.IpvSessionService;
-import uk.gov.di.ipv.core.library.service.UserIdentityService;
+import uk.gov.di.ipv.core.library.service.*;
```
Comment (blakeyp) [inline lambdas/initialise-ipv-session/src/test/java/uk/gov/di/ipv/core/initialiseipvsession/InitialiseIpvSessionHandlerTest.java]: We usually avoid wildcard imports - this should be a setting on your IDE that you can switch off

File: deploy/template.yaml [hunk 2]
```diff
@@ -488,6 +488,7 @@ Resources:
           POWERTOOLS_SERVICE_NAME: !Sub initialise-ipv-session-${Environment}
           IPV_SESSIONS_TABLE_NAME: !Select [ 1, !Split [ '/', !GetAtt SessionsTable.Arn ] ]
           SQS_AUDIT_EVENT_QUEUE_URL: !ImportValue AuditEventQueueUrl
+          IPV_CLIENT_OAUTH_SESSIONS_TABLE_NAME: !Ref ClientOAuthSessionsTable
```
Comment (blakeyp) [inline deploy/template.yaml]: Not a big deal but could drop the `IPV_` on this

File: lambdas/initialise-ipv-session/src/test/java/uk/gov/di/ipv/core/initialiseipvsession/InitialiseIpvSessionHandlerTest.java (hunk 1 above)
Comment (kerrr) [reply lambdas/initialise-ipv-session/src/test/java/uk/gov/di/ipv/core/initialiseipvsession/InitialiseIpvSessionHandlerTest.java]: +1

File: deploy/template.yaml (hunk 2 above)
Comment (kerrr) [reply deploy/template.yaml]: +1

Comment (lwebb-iig) [reply deploy/template.yaml]: Removed `IPV_` prefix

File: lambdas/initialise-ipv-session/src/test/java/uk/gov/di/ipv/core/initialiseipvsession/InitialiseIpvSessionHandlerTest.java (hunk 1 above)
Comment (lwebb-iig) [reply lambdas/initialise-ipv-session/src/test/java/uk/gov/di/ipv/core/initialiseipvsession/InitialiseIpvSessionHandlerTest.java]: Reverted to full imports

File: lib/src/main/java/uk/gov/di/ipv/core/library/service/ClientOAuthSessionDetailsService.java [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+            JWTClaimsSet claimsSet, String clientId) throws ParseException {
+        ClientOAuthSessionItem clientOAuthSessionItem = new ClientOAuthSessionItem();
+
+        clientOAuthSessionItem.setClientOAuthSessionId(SecureTokenHelper.generate());
+        clientOAuthSessionItem.setResponseType(claimsSet.getStringClaim("response_type"));
+        clientOAuthSessionItem.setClientId(clientId);
+        clientOAuthSessionItem.setRedirectUri(claimsSet.getStringClaim("redirect_uri"));
+        clientOAuthSessionItem.setState(claimsSet.getStringClaim("state"));
+        clientOAuthSessionItem.setUserId(claimsSet.getSubject());
+        clientOAuthSessionItem.setGovukSigninJourneyId("govuk_signin_journey_id");
```
Comment (TalatAydinGDS) [inline lib/src/main/java/uk/gov/di/ipv/core/library/service/ClientOAuthSessionDetailsService.java]: I think, should read that from claimsSet.getStringClaim("govuk_signin_journey_id") like generateClientSessionDetails()

Comment (lwebb-iig) [reply lib/src/main/java/uk/gov/di/ipv/core/library/service/ClientOAuthSessionDetailsService.java]: @TalatAydinGDS have changed as you've suggested & updated tests

**Review by TalatAydinGDS — APPROVED** [review]
Looks good to me.

**Pull request merged**